frontera/contrib/backends/redis_backend/__init__.py

```diff
@@ -0,0 +1,387 @@
+# -*- coding: utf-8 -*-
+from collections import Iterable
+from datetime import datetime
+from frontera.utils.url import parse_domain_from_url_fast
+from frontera import DistributedBackend
+from frontera.core.components import Metadata, Queue, States
+from frontera.core.models import Request
+from frontera.contrib.backends.partitioners import Crc32NamePartitioner
+from frontera.utils.misc import get_crc32
+from frontera.contrib.backends.remote.codecs.msgpack import Decoder, Encoder
+import logging
+from msgpack import packb, unpackb
+from redis import ConnectionPool, StrictRedis
+from time import time
+
+FIELD_CRAWL_AT = b'crawl_at'
+FIELD_CREATED_AT = b'created_at'
+FIELD_DEPTH = b'depth'
+FIELD_DOMAIN = b'domain'
+FIELD_DOMAIN_FINGERPRINT = b'domain_fingerprint'
+FIELD_ERROR = b'error'
+FIELD_FINGERPRINT = b'fingerprint'
+FIELD_NAME = b'name'
+FIELD_SCORE = b'score'
+FIELD_STATE = b'state'
+FIELD_STATUS_CODE = b'status_code'
+FIELD_URL = b'url'
+
+logging.getLogger('boto3.resources.action').setLevel(logging.WARNING)
+
+
+class RedisQueue(Queue):
+    MAX_SCORE = 1.0
+    MIN_SCORE = 0.0
+    SCORE_STEP = 0.01
+
+    def __init__(self, pool, partitions, delete_all_keys=False):
+        self._pool = pool
+        self._partitions = [i for i in range(0, partitions)]
+        self._partitioner = Crc32NamePartitioner(self._partitions)
+        self._logger = logging.getLogger("redis.queue")
+
+        if delete_all_keys:
+            connection = StrictRedis(connection_pool=self._pool)
+            connection.flushdb()
+
+        class DumbResponse:
+            pass
+
+        self._decoder = Decoder(Request, DumbResponse)
+        self._encoder = Encoder(Request)
+
+    def get_next_requests(self, max_n_requests, partition_id, **kwargs):
```

Comment:
The Redis backend is meant for storing metadata and not the actual crawled pages themselves. The limit is the size of the Redis install. As such it it meant for larger crawls. For smaller crawls you could use the Memory backend.

Both broad and deep crawling works well. My needs were primarily broad and since Redis has good random access speed, it is a good fit.

The maximum size of a managed Redis in Microsoft Azure is 53GB at the moment. That would hold a fair amount of metadata. AWS ElastiCache has a max Redis size of 3.55TB. A self-hosted Redis cluster could be bigger.